src/algebra/power_mod.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Given a monoid `M` and an element `g ∈ M` satisfying `gⁿ = 1`, we
+can interpret `gⁱ` for elements `i ∈ ℤ/n`.  We give several
+different Lean versions of this idea.  Some of them involve
+the type `zmod n`, which is Lean's implementation of `ℤ/n`.
+Others involve the relation `a ≡ b [MOD n]` on `ℕ`, or the
+relation `a ≡ b [ZMOD n]` on `ℤ`.
+-/
+
+import data.fintype algebra.group_power data.zmod.basic data.nat.modeq
+
+section pow_congr_monoid
+
+variables {n : ℕ} {M : Type*} [monoid M] {g : M} (hg : g ^ n = 1)
+include hg
+
+lemma pow_nat_mod (i : ℕ) : g ^ i = g ^ (i % n) :=
+calc
+  g ^ i = g ^ (i % (n : ℕ) + (n : ℕ) * (i / (n : ℕ))) :
+   by rw [nat.mod_add_div i (n : ℕ)]
+  ... = g ^ (i % (n : ℕ)) : by rw [pow_add, pow_mul, hg, one_pow, mul_one]
+
+lemma pow_nat_congr {i j : ℕ} (e : i ≡ j [MOD n]) : g ^ i = g ^ j :=
+begin
+  change i % n = j % n at e,
+  rw [pow_nat_mod hg i,pow_nat_mod hg j, e]
+end
+
+end pow_congr_monoid
+
+section pow_congr_group
+
+variables {n : ℕ} {G : Type*} [group G] {g : G} (hg : g ^ (n : ℕ) = 1)
+include hg
+
+lemma gpow_exponent (m : ℤ) : (g ^ m) ^ (n : ℕ) = 1 :=
+by rw [← gpow_coe_nat, ← gpow_mul, mul_comm,
+         gpow_mul, gpow_coe_nat, hg, one_gpow]
+
+lemma inv_exponent : (g⁻¹) ^ (n : ℕ) = 1 := by rw [inv_pow,hg,one_inv]
+
+lemma gpow_nat_congr {i j : ℕ} (e : i ≡ j [MOD n]) : g ^ i = g ^ j :=
+begin
+  change i % n = j % n at e,
+  rw [pow_nat_mod hg i, pow_nat_mod hg j, e]
+end
+
+lemma gpow_int_mod (i : ℤ) : g ^ i = g ^ (i % n) :=
+by rw[← congr_arg ((^) g) (int.mod_add_div i n), gpow_add, gpow_mul,
+      gpow_coe_nat, hg, one_gpow, mul_one]
+
+lemma gpow_int_mod' (h : n > 0) (i : ℤ) : g ^ i = g ^ (i % n).nat_abs :=
+begin
+ have := int.mod_nonneg i (ne_of_gt (int.coe_nat_pos.mpr h)),
```

Comment:
Another way to write down the same proof would be to start with
```lean
  suffices : ((i % n).nat_abs : ℤ) = i % n,
  { ... },
  ...
```
Both the `...` can probably be proven with a relatively short call to `simp` or `simpa`.